### PR TITLE
Bugfix/973 epoch with stimsets and multiple sweeps

### LINKED ISSUE #827: EpochTest6 fails with more delays
When I apply

```diff
--- a/Packages/Testing-MIES/UTF_Epochs.ipf
+++ b/Packages/Testing-MIES/UTF_Epochs.ipf
@@ -66,6 +66,13 @@ static Function AcquireData(s, devices, stimSetName1, stimSetName2[, dDAQ, oodDA

         PGC_SetAndActivateControl(device, "Check_DataAcq1_DistribDaq", val = dDAQ)
         PGC_SetAndActivateControl(device, "Check_DataAcq1_dDAQOptOv", val = oodDAQ)
+
+        if(dDAQ)
+            PGC_SetAndActivateControl(device, "Setvar_DataAcq_dDAQDelay", val = 50)
+        elseif(oodDAQ)
+            PGC_SetAndActivateControl(device, "setvar_DataAcq_dDAQOptOvPost", val = 75)
+            PGC_SetAndActivateControl(device, "setvar_DataAcq_dDAQOptOvPre", val = 35)
+        endif

         PASS()
     endfor
```

and run `UTF_Epochs.ipf` I get the following failure:

```
  Entering reentry "EP_EpochTest6_REENTRY"
  0: is false.  Assertion "CHECK(endT[epochNr] <= endInt)" failed in line 189 (408), procedure "UTF_Epochs.ipf"
  0: is false.  Assertion "REQUIRE(startT[epochNr] >= 0 && endT[epochNr] <= endInt + OODDAQ_PRECISION)" failed in line 185 (408), procedure "UTF_Epochs.ipf"
  Leaving test case "EP_EpochTest6:ITC"
```

Investigate what is at fault here.

### PR DESCRIPTION
**Pull request by t-b**
Close #827

### CODE REVIEW

**Comment by t-b** [comment]
@timjarsky This is the screenshot in the documentation.
grafik

**Review by MichaelHuth — APPROVED** [review]
Looks good.

I was a bit unsure if it is also true for specific cases, like flipped oodDAQ, but the stimset extension for multiple stimsets with different length should be independent of that.

**Comment by t-b** [comment]
This needs some more fixes, taking it back.

**Comment by t-b** [comment]
@MichaelHuth Can you have a look?

**Comment by t-b** [comment]
@timjarsky FYI: This now fixes all known epoch info bugs. The ones I fixed today are rather obscure though.

File: Packages/MIES/MIES_DataConfigurator.ipf [hunk 1]
```diff
@@ -2347,11 +2347,13 @@ static Function DC_AddEpochsFromStimSetNote(panelTitle, channel, stimset, stimse
 						// shift all flipped pulse intervalls by pulseDuration to the left, except the rightmost with pulseNr 0
 						if(!pulseNr)
 							subEpochBegin = epochEnd - startTimes[0] - pulseDuration
-							subEpochEnd = subEpochBegin + pulseDuration
+							// assign left over time after the last pulse to that pulse
+							subEpochEnd = epochEnd
 						else
-							subEpochEnd = epochEnd - startTimes[pulseNr - 1] - pulseDuration
+							subEpochEnd = epochEnd - startTimes[limit(pulseNr - 1, 0, inf)] - pulseDuration
```
Comment (MichaelHuth) [inline Packages/MIES/MIES_DataConfigurator.ipf:2353]: Why is the limit added here?
The else part of the if check can only be run with pulseNr > 0.

File: Packages/MIES/MIES_DataConfigurator.ipf [hunk 2]
```diff
@@ -2376,9 +2378,15 @@ static Function DC_AddEpochsFromStimSetNote(panelTitle, channel, stimset, stimse
...
 						// active
 						subsubEpochBegin = subEpochBegin
-						// we never have a trailing baseline after the last pulse in the epoch
-						// to avoid decimation errors we assign all of the left over time to pulse active
-						subsubEpochEnd = (pulseNr == (flipping ? 0 : numPulses - 1)) ? subEpochEnd : (subEpochBegin + pulseDuration)
+
+						// normally we never have a trailing baseline with pulse train
+						// but if poisson distribution and flipping is combined we can have it
+						// to avoid decimation errors we assign all of the left over time to pulse active if we are not in this special case
+						if(!poissonDistribution && (pulseNr == (flipping ? 0 : numPulses - 1)))
```
Comment (MichaelHuth) [inline Packages/MIES/MIES_DataConfigurator.ipf:2385]: I think the if check somehow does not reflect the comment above as it does
subsubEpochEnd = subEpochBegin + pulseDuration
also for the (poissonDistribution + non-flipped) case.

**Review by MichaelHuth — COMMENTED** [review]
I have some quetions.

File: Packages/MIES/MIES_DataConfigurator.ipf (hunk 1 above)
Comment (t-b) [reply Packages/MIES/MIES_DataConfigurator.ipf:2353]: Good point. Will remove.

File: Packages/MIES/MIES_DataConfigurator.ipf (hunk 2 above)
Comment (t-b) [reply Packages/MIES/MIES_DataConfigurator.ipf:2385]: Thx. Fixed.

**Review by MichaelHuth — APPROVED** [review]

**Comment by timjarsky** [comment]
@t-b can we test this together the next time we chat - I want to take advantage of the special plotting tools to see if the epochs handle shuffle pulses correction

**Review by timjarsky — APPROVED** [review]

**Pull request merged**